src/ansi.rs

```diff
@@ -0,0 +1,1943 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// This module was originally part of the `alacritty_terminal` crate, which is
+// licensed under the Apache License, Version 2.0 and is part of the Alacritty
+// project (https://github.com/alacritty/alacritty).
+
+//! ANSI Terminal Stream Parsing.
+
+extern crate alloc;
+
+use alloc::borrow::ToOwned;
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+
+use core::convert::TryFrom;
+use core::fmt::Write;
+use core::time::Duration;
+use core::{iter, str};
+
+use log::{debug, trace};
+
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{Params, ParamsIter};
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct Hyperlink {
+    /// Identifier for the given hyperlink.
+    pub id: Option<String>,
+    /// Resource identifier of the hyperlink.
+    pub uri: String,
+}
+
+#[derive(Debug, Eq, PartialEq, Copy, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Rgb {
```

Comment:
The `Rgb` struct in `alacritty_terminal` provides additional functions (`luminance` and `contrast`) and implements additional traits (`Mul<f32>` and `Add<Rgb>`).

In the initial version of making Alacritty use VTE ANSI parser (https://github.com/alacritty/alacritty/pull/6896), I just convert between the Alacritty and VTE `Rgb` structs. By adding those functions in VTE would allow us to get rid of the type from `alacritty_terminal` but will still require us to have a `RgbShim` type so that we can implement the custom deserialize for ref tests and serde replace where serialization is required.

Which approach would you guys suggest to take?

Reply:
Why not just move the functions to VTE? If Alacritty requires it, there's a good chance other emulators might too.

Reply:
I have moved that functionality to VTE (https://github.com/alacritty/vte/pull/85/commits/95132a6ba9b23228d2b5c0511609c1b59be97dba and https://github.com/alacritty/vte/pull/85/commits/187cdc8a3bb7d12794bb918ee8acad23e236a3e7). I have also updated Alacritty to make use of the VTE provided RGB type under the hood (https://github.com/alacritty/alacritty/pull/6896/commits/37c22c800187cf78551a2756b5facb3f940fbb0d).